python/hail/matrixtable.py

```diff
@@ -657,78 +689,161 @@ def entry(self):
         return self._entry
 
-    @typecheck_method(keys=oneof(str, Expression))
-    def key_cols_by(self, *keys) -> 'MatrixTable':
+    @typecheck_method(keys=oneof(str, Expression),
+                      named_keys=Expression)
+    def key_cols_by(self, *keys, **named_keys) -> 'MatrixTable':
         """Key columns by a new set of fields.
 
+        See :meth:`.MatrixTable.key_rows_by` for more examples.
+
         Parameters
         ----------
-        keys : varargs of :obj:`str`
-            Column fields to key by.
+        keys : varargs of :obj:`str` or :class:`.Expression`.
+            Row fields to key by.
+        named_keys : keyword args of :class:`.Expression`.
+            Row fields to key by.
         Returns
         -------
         :class:`.MatrixTable`
         """
-        str_keys = []
-        for k in keys:
-            if isinstance(k, Expression):
-                if k not in self._fields_inverse:
-                    raise ExpressionException("'key_cols_by' permits only top-level fields of the matrix table")
-                elif k._indices != self._col_indices:
-                    raise ExpressionException("key_cols_by' expects column fields, found index {}"
-                                              .format(list(k._indices.axes)))
-                str_keys.append(self._fields_inverse[k])
-            else:
-                if k not in self._fields:
-                    raise LookupError(get_nice_field_error(self, k))
-                if not self._fields[k]._indices == self._col_indices:
-                    raise ValueError("'{}' is not a column field".format(k))
-                str_keys.append(k)
-        return MatrixTable(self._jvds.keyColsBy(str_keys))
+        key_fields = get_select_exprs("MatrixTable.key_cols_by",
+                                keys, named_keys, self._col_indices,
+                                protect_keys=False)
+        return self._select_cols("MatrixTable.key_cols_by",
+                                 key_struct=hl.struct(**key_fields))
 
-    @typecheck_method(keys=oneof(str, Expression),
-                      partition_key=nullable(oneof(oneof(str, Expression), sequenceof(oneof(str, Expression)))))
-    def key_rows_by(self, *keys, partition_key=None) -> 'MatrixTable':
-        """Key rows by a new set of fields.
+    def _key_rows_by(self, caller, pk_dict, rest_of_keys_dict):
+
+        for k in rest_of_keys_dict.keys():
+            if k in pk_dict:
+                raise ValueError("cannot have two row key fields with the same name: '{}'".format(k))
+
+        key_fields = dict(pk_dict, **rest_of_keys_dict)
+
+        return self._select_rows(caller, key_struct=hl.struct(**key_fields), pk_size=len(pk_dict))
+
+    @typecheck_method(partition_key=oneof(str, sequenceof(str)),
+                      keys=oneof(str, Expression),
+                      named_keys=Expression)
+    def partition_rows_by(self, partition_key, *keys, **named_keys) -> 'MatrixTable':
```

Comment:
I would prefer an interface where this just takes `*partition_key`. I'll explain my thoughts today.

Reply:
yeah, this was me taking the lazy way out >.> let's talk?